libp2p/protocols/pubsub/gossipsub.nim

```diff
@@ -21,40 +21,137 @@ import ./pubsub,
+  IWantPeerBudget = 25 # 25 messages per second ( reset every heartbeat )
+  IHavePeerBudget = 10
 
 type
+  TopicInfo* = object
+    # gossip 1.1 related
+    graftTime: Moment
+    meshTime: Duration
+    inMesh: bool
+    meshMessageDeliveriesActive: bool
+    firstMessageDeliveries: float64
+    meshMessageDeliveries: float64
+    meshFailurePenalty: float64
+    invalidMessageDeliveries: float64
+
+  TopicParams* = object
+    topicWeight*: float64
+
+    # p1
+    timeInMeshWeight*: float64
+    timeInMeshQuantum*: Duration
+    timeInMeshCap*: float64
+
+    # p2
+    firstMessageDeliveriesWeight*: float64
+    firstMessageDeliveriesDecay*: float64
+    firstMessageDeliveriesCap*: float64
+
+    # p3
+    meshMessageDeliveriesWeight*: float64
+    meshMessageDeliveriesDecay*: float64
+    meshMessageDeliveriesThreshold*: float64
+    meshMessageDeliveriesCap*: float64
+    meshMessageDeliveriesActivation*: Duration
+    meshMessageDeliveriesWindow*: Duration
+
+    # p3b
+    meshFailurePenaltyWeight*: float64
+    meshFailurePenaltyDecay*: float64
+
+    # p4
+    invalidMessageDeliveriesWeight*: float64
+    invalidMessageDeliveriesDecay*: float64
+
+  PeerStats* = object
+    topicInfos*: Table[string, TopicInfo]
+    expire*: Moment # updated on disconnect, to retain scores until expire
+
+  GossipSubParams* = object
+    explicit: bool
+    pruneBackoff*: Duration
+    floodPublish*: bool
+    gossipFactor*: float64
+    dScore*: int
+    dOut*: int
+    dLazy*: int
+
+    gossipThreshold*: float64
+    publishThreshold*: float64
+    graylistThreshold*: float64
+    acceptPXThreshold*: float64
+    opportunisticGraftThreshold*: float64
+    decayInterval*: Duration
+    decayToZero*: float64
+    retainScore*: Duration
+
+    appSpecificWeight*: float64
+    ipColocationFactorWeight*: float64
+    ipColocationFactorThreshold*: float64
+    behaviourPenaltyWeight*: float64
+    behaviourPenaltyDecay*: float64
+
+    directPeers*: seq[PeerId]
+
   GossipSub* = ref object of FloodSub
     mesh*: PeerTable                           # peers that we send messages to when we are subscribed to the topic
     fanout*: PeerTable                         # peers that we send messages to when we're not subscribed to the topic
     gossipsub*: PeerTable                      # peers that are subscribed to a topic
+    explicit*: PeerTable                       # directpeers that we keep alive explicitly
+    backingOff*: Table[PeerID, Moment]         # explicit (always connected/forward) peers
     lastFanoutPubSub*: Table[string, Moment]   # last publish time for fanout topics
     gossip*: Table[string, seq[ControlIHave]]  # pending gossip
     control*: Table[string, ControlMessage]    # pending control messages
     mcache*: MCache                            # messages cache
     heartbeatFut: Future[void]                 # cancellation future for heartbeat interval
     heartbeatRunning: bool
 
+    peerStats: Table[PubSubPeer, PeerStats]
```

Comment:
why not add a `PeerStats` instance to `PubSubPeer` instead?

Reply:
The issue is that PeerStats outlives PubSubPeer, in order to keep stats alive for X time even if peers disconnect.